src/deno_json/mod.rs

```diff
@@ -581,115 +612,69 @@ fn decorate_tasks_json(
+        CONFIG_FILE_NAMES
+          .iter()
+          .copied()
+          .chain(additional_config_file_names.iter().copied())
+          .collect::<Vec<_>>(),
+      )
     }
   }
 
   pub fn discover_from(
+    fs: &dyn DenoConfigFs,
     start: &Path,
-    checked: &mut HashSet<PathBuf>,
-    additional_config_file_names: Option<&Vec<&str>>,
-    parse_options: &ParseOptions,
-  ) -> Result<Option<ConfigFile>, AnyError> {
-    fn is_skippable_err(e: &AnyError) -> bool {
-      if let Some(ioerr) = e.downcast_ref::<std::io::Error>() {
-        use std::io::ErrorKind::*;
-        match ioerr.kind() {
-          InvalidInput | PermissionDenied | NotFound => {
-            // ok keep going
-            true
-          }
-          _ => {
-            const NOT_A_DIRECTORY: i32 = 20;
-            cfg!(unix) && ioerr.raw_os_error() == Some(NOT_A_DIRECTORY)
-          }
-        }
+    additional_config_file_names: &[&str],
+    parse_options: &ConfigParseOptions,
+  ) -> Result<Option<ConfigFile>, ConfigFileReadError> {
+    fn is_skippable_err(e: &ConfigFileReadError) -> bool {
+      if let ConfigFileReadError::FailedReading { source: ioerr, .. } = e {
+        is_skippable_io_error(ioerr)
       } else {
         false
       }
     }
 
-    /// Filenames that Deno will recognize when discovering config.
-    const CONFIG_FILE_NAMES: [&str; 2] = ["deno.json", "deno.jsonc"];
     let config_file_names =
-      if let Some(additional) = additional_config_file_names {
-        CONFIG_FILE_NAMES
-          .into_iter()
-          .chain(additional.to_vec())
-          .collect::<Vec<_>>()
-      } else {
-        CONFIG_FILE_NAMES.to_vec()
-      };
-
-    // todo(dsherret): in the future, we should force all callers
-    // to provide a resolved path
-    let start = if start.is_absolute() {
-      Cow::Borrowed(start)
-    } else {
-      Cow::Owned(std::env::current_dir()?.join(start))
-    };
+      Self::resolve_config_file_names(additional_config_file_names);
 
+    debug_assert!(start.is_absolute());
     for ancestor in start.ancestors() {
-      if checked.insert(ancestor.to_path_buf()) {
-        for config_filename in &config_file_names {
-          let f = ancestor.join(config_filename);
-          match ConfigFile::read(&f, parse_options) {
-            Ok(cf) => {
-              log::debug!("Config file found at '{}'", f.display());
-              return Ok(Some(cf));
-            }
-            Err(e) if is_skippable_err(&e) => {
-              // ok, keep going
-            }
-            Err(e) => {
-              return Err(e);
-            }
+      for config_filename in config_file_names.iter() {
+        let f = ancestor.join(config_filename);
+        match ConfigFile::read(fs, &f, parse_options) {
+          Ok(cf) => {
+            log::debug!("Config file found at '{}'", f.display());
+            return Ok(Some(cf));
+          }
+          Err(e) if is_skippable_err(&e) => {
+            // ok, keep going
+          }
```

Comment:
Have you benched if this approach is faster than `f.exists()`?

Reply:
No, but I feel like it should be. I will look into it.